src/main/scala/org/phenoscape/kb/Graph.scala

```diff
@@ -123,40 +122,53 @@ object Graph {
          $queryPattern
          FILTER(?subsumer != $owlThing)
        }
-          """
-
-      val futurePairs = App.executeSPARQLQueryString(
-        query.text,
-        qs => {
-          val term = qs.getResource("term").getURI
-          val relation = qs.getResource("relation").getURI
-          val subsumer = qs.getResource("subsumer").getURI
-          (term, relation, subsumer)
+       """
+
+    val futurePairs = App.executeSPARQLQueryString(
+      query.text,
+      qs => {
+        val term = qs.getResource("term").getURI
+        val relation = qs.getResource("relation").getURI
+        val subsumer = qs.getResource("subsumer").getURI
+        (term, relation, subsumer)
+      }
+    )
+
+    val termSubsumerSeq = for {
+      pairs <- futurePairs
+    } yield {
+      val termSubsumerPairs = pairs.map(p => (p._1 -> (p._2, p._3)))
+
+      termSubsumerPairs.map { case (term, pairs) =>
+        val virtualTermIRI = pairs._1 match {
+          case "http://www.w3.org/2000/01/rdf-schema#subClassOf" => IRI.create(pairs._2)
+          case _                                                 => RelationalTerm(IRI.create(pairs._1), IRI.create(pairs._2)).iri
+
         }
-      )
+        Map(IRI.create(term) -> virtualTermIRI)
+      }.flatten
+    }
+    termSubsumerSeq
+  }
+
+  def ancestorMatrix(terms: Set[IRI], relations: Set[IRI], pathOpt: Option[Path]): Future[AncestorMatrix] =
+    if (terms.isEmpty) Future.successful(AncestorMatrix(""))
+    else {
+      val termSubsumerPairsFut = getTermSubsumerPairs(terms, relations, pathOpt)
       for {
-        pairs <- futurePairs
+        termSubsumerPairs <- termSubsumerPairsFut
       } yield {
         val termsSequence = terms.map(_.toString).toSeq.sorted
```

Comment:
I think I would use something like `sortBy(_.toString)` instead of converting all to strings, because later on line 168 you have to create an IRI again.